storage/pebble/operation/common.go

```diff
@@ -10,14 +12,71 @@ import (
 	"github.com/onflow/flow-go/storage"
 )
 
+type ReaderBatchWriter struct {
+	db        *pebble.DB
+	batch     *pebble.Batch
+	callbacks []func()
+}
```

Comment:
badger supports transactions. However, pebble doesn't. 
pebble only supports atomic batch updates. This is the key difference. 

With badger transaction, we have a lot operations that create a badger transaction, then reads some data, and update some data depending on the result of the reads, and if the transaction is committed successfully, then we also use callback to update cache. 

In order to implement a similar behavior as badger transaction in pebble, I create this ReaderBatchWriter struct. It contains the batch for updates. If we need to read something within the transaction, then the `db` is be used to read data. 

When reading the data, there are chances where we need to take into consideration the writes that has not been committed, this is why I made the `IndexedBatch` method to return the Batch object, which holds all the write set. 

And it also contains the callback that allows us to update cache after successfully committing the batch.

Reply:
## ⚠️ 

I think this is a highly unsafe construction for this refactoring, because it hides conceptually breaking changes for the business logic at a very low level. Further explanation:
* Previously, reads and writes within the same transaction all operated atomically on a snapshot of the data base.  For example, I could read the _latest_ finalized bock within a transaction and decide to write something depending on what I found for the _latest_ finalized bock.
* The `ReaderBatchWriter` you are proposing here breaks this. I _can_ read the _latest_ finalized block and decide what I want to write in response, but by the time the write hits the data base, the latest finalized block might have already changed. While the `ReaderBatchWriter` exposes similar functionality as a badger transaction (reading and writing) it doesn't provide atomicity guarantees of reads at all. 
* what we need to do is to go through _every_ piece of business logic that previously used a badger transaction, confirm that we can asynchronously do the reads upfront and then change the code. 
* I think it would be much safer to not provide `ReaderBatchWriter` and force ourselves to re-write every single occurrence where we previously used a badger transaction. I think that it would be much much more likely for us to find edge cases when we force ourselves to look at each occurrence of badger transaction with reads and refactor it. It is important to be aware that human brains are just not very good at _exhaustively_ finding _all_ occurrences of `ReaderBatchWriter`  changes the guarantees for component `A`, which in turn changes guarantees for component `B` and `C`. 

From my perspective, we are doing a low-level change like this, so that the compiler is happy with the higher-level logic, yet guarantees for the higher-level logic have changed and we relying on humans to identify those areas where something could go wrong. I think an approach like this has disproportional risks to overlook bugs that we cannot afford in a project like Flow.